prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -588,6 +592,28 @@ def build_transmission(method="sub2line", kwargs={"rounding": 3}):
     substations = filter_substations_with_zero_lines(hifld_substations)
     check_for_location_conflicts(substations)
 
+    # Filter out keyword arguments for get_mst_edges function
+    get_mst_kwargs = dict()
+    for keyword in {
+        "island_size_lower_bound",
+        "island_size_upper_bound",
+        "state_neighbor",
+        "min_dist_method",
+        "cost_metric",
+        "memory_efficient",
+        "kdtree_kwargs",
+    }:
+        if keyword in kwargs:
```

Comment:
If we stick with the kwargs pattern, I would change the signature to be the following, since the dict is mutable, and we are also modifying it here.

```
def build_transmission(method="sub2line", **kwargs):
 ```

Reply:
Maybe we need some background story from @rouille or @danielolsen on the current implementation?

Reply:
@jon-hagg If I'm understanding correctly, the idea of the design would be in function definition, any `kwarg` hits the lowest level function, it should be declared explicitly and everything else can be collapsed in `**kwargs`? Suppose we have the main user facing function `funcA(kwarg_a, **kwargs)` and it calls `funcB` and `funcC`, which take different entries in `kwargs` as keyword arguments, how do will filter out those arguments during the function call so that it won't throw `TypeError: funcB() got an unexpected keyword argument 'xxx'`?

Reply:
I guess there are two ways I see to handle it. The current way pops off each argument before passing to lower level functions with explicit keyword arguments defined. Another way would be having lower level functions just take arbitrary kwargs (in your example, it would be `funcB(**kwargs)` and same for `funcC`), and pop the ones they are interested in. 

My original comment was mainly about the mutable default argument, which is a pattern we avoid in general. I don't have a strong opinion on whether to declare it as `def foo(kwargs=None)` or `def foo(**kwargs)`.

Reply:
There's some backstory here: https://github.com/Breakthrough-Energy/PreREISE/pull/217#discussion_r705552095

The main design goal I think was to enable the user to have a lot of flexibility from the highest-level `build_transmission` function, to be able to specify parameters for several lower-level functions, without cluttering the `build_transmission` docstring too bad. We do something similar in PostREISE for the map functions that can pass keyword arguments down to the lower-level function that builds the background states map.

If we stick with one big dict in the top-level function, then it seems the design candidates for how to handle it are:
a. Pass all kwargs to all lower-level functions. Pro: simplicity of code. Con: it's less clear which keyword arguments will be interpreted by each lower-level function.
b. Within the highest-level function, split the one big dictionary into smaller dictionaries, each of which is passed to its corresponding lower-level function. Pro: clarity of lower-level function docstrings. Con: the higher-level function gets a bit more complex.

Reply:
The thing is, `KDTree` itself is not implemented by us, i.e. we can't pass arbitrary `kwargs` to it at the lower level function that creates the `KDTree`. Instead, we still need to filter out what `kwargs` should go into `KDTree` creation and `KDTree.query` at the end, right? Filtering at the top level function will give us a clean doc string structure for all lower level functions.

In any case, we will need convergence in the options we have been discussing on this track:
1. The current implementation: the user facing function takes `**kwargs`, then we filter which `kwargs` to pass to lower level functions before the function call. Lower level functions will have `kwargs` defined explicitly.
2. Use one big dict `kwargs` in the top-level function. Pass all `kwargs` to all lower level functions. All lower level functions will take `**kwargs`.
3. Use several smaller dicts in the top-level function. Pass specific dict to lower level functions.
4. Top-level function is the same as 2. But use classes internally, pass `kwargs` upon class instance creation.
5. All of class approach.

Reply:
I like option 1 the best, and option 3 second-best.

Reply:
I like 1 the best

Reply:
Done with [77d6ce7](https://github.com/Breakthrough-Energy/PreREISE/pull/228/commits/77d6ce7c94e54acced97d04da4d83870683c32da)
I did a bunch of refactor to make the `kwargs` passing logic consistent throughout the module based on option 1:
- A new function `filter_islands_and_connect_with_mst` is added following @danielolsen 's suggestion in this [thread](https://github.com/Breakthrough-Energy/PreREISE/pull/228#discussion_r718054690), which keeps `build_transmission` as purely a collection of calls to low/medium-level functions, with no lowest-level logic.
- `get_mst_edges` is moved to `topology` module.
- According to option 1, top level user facing function `build_transmission` only takes `**kwargs`, which is further split into two subsets feeding into `map_lines_to_substations_using_coords` and `filter_islands_and_connect_with_mst` respectively. The former function is refactored accordingly to reflect the change.